### PR TITLE
File-level reflection, C++ and F# code generation, golden tests.

### PR DESCRIPTION
**Pull request by dkorolev**
Hi @mzhurovich,

Before I dive back into polymorphics, here's an attempt to:
* move reflection / schema generation one level higher, to source file / group of structures level, and
* add F# support.

Thanks,
Dima

### CODE REVIEW

File: TypeSystem/Reflection/golden/smoke_test_struct.json [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+    ],
+    [
+      9209518734145112000,
+      {
+        "type_id": 9209518734145112000,
+        "name": "Primitives",
+        "super_type_id": 0,
+        "fields": [
+          [
+            9000000000000000000,
```
Comment (mzhurovich) [inline TypeSystem/Reflection/golden/smoke_test_struct.json]: Something went wrong with the primitive TypeIDs :(

Comment (mzhurovich) [inline TypeSystem/Reflection/golden/smoke_test_struct.json]: Test fails on both - OS X and Ubuntu - with correct TypeIDs in `Actual`.
Looks like you just committed wrong file.

File: TypeSystem/Reflection/schema.h [hunk 2]
```diff
@@ -244,107 +297,106 @@ struct StructSchema {
...
     const uint64_t type_prefix = TypePrefix(type_id);
-    if (type_prefix == TYPEID_STRUCT_PREFIX) {
-      if (schema_.structs.count(type_id) == 0u) {
-        return "UNKNOWN_STRUCT_" + bricks::strings::ToString(type_id);
+    if (type_prefix == TYPEID_BASIC_PREFIX) {
+      return primitive_types_.PrimitiveTypeName(language, type_id);
+    } else if (type_prefix == TYPEID_STRUCT_PREFIX) {
+      const auto cit = schema_.structs.find(type_id);
+      if (cit == schema_.structs.end()) {
+        return "UNKNOWN_STRUCT__" + bricks::strings::ToString(type_id);
```
Comment (mzhurovich) [inline TypeSystem/Reflection/schema.h]: Double underscore here?

File: TypeSystem/Reflection/schema.h [hunk 3]
```diff
@@ -244,107 +297,106 @@ struct StructSchema {
...
+  std::string DescribeMap(const Language::CPP& language, const TypeInfo& type_info) const {
+    assert(type_info.included_types.size() == 2u);
+    return "std::map<" + TypePrintName(language, type_info.included_types[0]) + ", " +
+           TypePrintName(language, type_info.included_types[1]) + '>';
+  }
+  std::string DescribeMap(const Language::FSharp& language, const TypeInfo& type_info) const {
     assert(type_info.included_types.size() == 2u);
-    return "std::map<" + CppType(type_info.included_types[0]) + ", " + CppType(type_info.included_types[1]) +
-           '>';
+    return "UNSUPPORTED_MAP<" + TypePrintName(language, type_info.included_types[0]) + ", " +
```
Comment (mzhurovich) [inline TypeSystem/Reflection/schema.h]: `Dictionary` from `System.Collections.Generic`?
https://en.wikibooks.org/wiki/F_Sharp_Programming/Mutable_Collections#HashSet.3C.27T.3E.2C_and_Dictionary.3C.27TKey.2C_.27TValue.3E_Classes

File: TypeSystem/Reflection/golden/smoke_test_struct.cc [hunk 4]
```diff
@@ -0,0 +1,32 @@
+// g++ -c -std=c++11 current.cc
+
+#include <string>
+#include <cstdint>
+#include <vector>
+#include <map>
+#include <utility>
```
Comment (mzhurovich) [inline TypeSystem/Reflection/golden/smoke_test_struct.cc]: Technically speaking we also need Current includes here :-|

File: TypeSystem/Reflection/schema.h (hunk 3 above)
Comment (dkorolev) [inline TypeSystem/Reflection/schema.h]: Never tested it within a discriminated union (may need `new` of some sort). Let's leave it as is now, our F# bindings would not need any per-entry dictionaries yet. Adding a `TODO(dkorolev)`.

File: TypeSystem/Reflection/schema.h (hunk 2 above)
Comment (dkorolev) [inline TypeSystem/Reflection/schema.h]: Fixed.

File: TypeSystem/Reflection/golden/smoke_test_struct.cc (hunk 4 above)
Comment (dkorolev) [inline TypeSystem/Reflection/golden/smoke_test_struct.cc]: Done. :trollface:

File: TypeSystem/Reflection/golden/smoke_test_struct.json (hunk 1 above)
Comment (dkorolev) [inline TypeSystem/Reflection/golden/smoke_test_struct.json]: AUCH.

It's the pretty-printing. Both `node` and `jq` "round" numbers to less than 64 bits. :-(

`json_pp` did it. 0e1f037

**Comment by dkorolev** [comment]
@mzhurovich Fixed all, and you're right about type IDs being the wrong file committed from my end. PTAL.

**Comment by mzhurovich** [comment]
@dkorolev :+1:

**Pull request merged**